File: java/src/main/java/UnitTestingInterviewQuestions.java
import models.Animal;
import helpers.TemperatureConstantGetter;
import models.ItsFarTooHotException;

public class UnitTestingInterviewQuestions {

    private final TemperatureConstantGetter temperatureConstantGetter;

    public UnitTestingInterviewQuestions(TemperatureConstantGetter temperatureConstantGetter) {
        this.temperatureConstantGetter = temperatureConstantGetter;
    }

    public boolean isItTooHot(int temperature, int humidity, Animal animal) throws ItsFarTooHotException {
        var hotnessRating = this.generateHotnessRating(temperature, humidity);

        if(hotnessRating > 100) {
            throw new ItsFarTooHotException();
        }

        return switch (animal) {
            case LION -> hotnessRating > 70;
            case POLAR_BEAR -> hotnessRating > 10;
            case CHINCHILLA -> hotnessRating > 30;
        };
    }

    int generateHotnessRating(int temperature, int humidity) {
        // TODO: Some additional maths that makes this function very long and complex
        var humidityRating = calculateHumidityRating(humidity);
        return (temperature * humidityRating) / this.temperatureConstantGetter.getHotnessConstant();
    }

    private int calculateHumidityRating(int humidity) {
        // TODO: More complex maths
        return humidity * this.temperatureConstantGetter.getHumidityConstant();
    }
}

File: cpp/unit_testing_interview_questions.cpp
#include "unit_testing_interview_questions.h"

using namespace testing_questions;

UnitTestingInterviewQuestions::UnitTestingInterviewQuestions(std::shared_ptr<TemperatureConstantGetter> temperature_constant_getter)
    : temperature_constant_getter_{temperature_constant_getter}
{}

bool UnitTestingInterviewQuestions::isItTooHot(int temperature, int humidity, Animal animal) const {
    auto hotness_rating = generateHotnessRating(temperature, humidity);
    if(hotness_rating > 100) {
        throw ItsFarTooHotException("Too hot for animal");
    }

    switch(animal) {
    case Animal::LION:
        return hotness_rating > 70;
    case Animal::POLAR_BEAR:
        return hotness_rating > 10;
    case Animal::CHINCHILLA:
        return hotness_rating > 30;
    }
    throw ItsFarTooHotException("Invalid animal");
}

int UnitTestingInterviewQuestions::generateHotnessRating(int temperature, int humidity) const {
    auto humidity_rating = calculateHumidityRating(humidity);
    return (temperature * humidity_rating) / temperature_constant_getter_->getHotnessConstant();
}

int UnitTestingInterviewQuestions::calculateHumidityRating(int humidity) const {
    // TODO: More complex maths
    return humidity * temperature_constant_getter_->getHumidityConstant();
}
